lib/telemetry_registry.ex

```diff
@@ -0,0 +1,253 @@
+
+  @typedoc """
+  A description of what the event represents and when it is emitted.
+  """
+  @type event_description() :: :telemetry_registry.event_description()
+
+  @typedoc """
+  A string representation of the measurements emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include a comment on which unit a value is in. The objective
+  is to inform users.
+  """
+  @type event_measurements() :: :telemetry_registry.event_measurements()
+
+  @typedoc """
+  A string representation of the metadata emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include comments or example values. The objective
+  is to inform users what is available.
+  """
+  @type event_metadata() :: :telemetry_registry.event_metadata()
+
+  @typedoc """
+  A map of event definition meta for an event containing the event, measurements, and metadata descriptions
+  if the event was declared with an event definition. Otherwise, this value will be an empty map.
+  """
+  @type event_meta() :: :telemetry_registry.event_meta()
+
+  @typedoc """
+  A list of spannable events known to the registry in the format of `{event_prefix, event_suffixes}`. For
+  example, given events `[:my_app, :request, :start], [:my_app, :request, :stop], [:my_app, :request, :exception]`
+  a span can be created from the `:start` -> `:stop` or the `:start` -> `:exception` events. These are aggregated
+  as a spannable event `{[:my_app, :request], [:start, :stop, :exception]}`.
+  """
+  @type spannable_event() :: :telemetry_registry.spannable_event()
+
+
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__), only: [format_telemetry_docs: 1, telemetry_event: 1]
+
+      Module.register_attribute(__MODULE__, :telemetry_event,
+        accumulate: true,
+        persist: true
+      )
+
+      Module.register_attribute(__MODULE__, :telemetry_docs,
+        accumulate: true,
+        persist: true
+      )
+    end
+  end
+
+  @doc """
+  Declares a telemetry event. Accepts a telemetry event definition or telemetry event name. A full
+  definition is required for complete documentation and is highly encouraged.
+  """
+  defmacro telemetry_event(event) do
+    docstring =
+      case event do
+        event when is_list(event) ->
+          """
+          * `#{inspect(event)}`
+
+          """
+
+        {:%{}, _, event} ->
+          """
+          * `#{inspect(event[:event])}`
+            * Description: #{event[:description]}
+            * Measurements: `#{event[:measurements]}`
+            * Metadata: `#{event[:metadata]}`
+
+          """
+      end
+
+    quote do
+      @telemetry_event unquote(event)
+      @telemetry_docs unquote(docstring)
+    end
+  end
+
+  @doc """
+  Formats telemetry definitions declared in the module for inclusion in documentation.
+
+  Example
+
+  ```
+  ## Telemetry
+
+  #{format_telemetry_docs(@telemetry_docs)}
```

Comment:
You need to escape the interpolation here OR use the `~S"""` sigil to disable them. :)

```suggestion
  \#{format_telemetry_docs(@telemetry_docs)}
```